### PR TITLE
feat: added socket buffers size parameters to config

### LINKED ISSUE #24104: Make SO_SNDBUF & SO_RCVBUF configurable in NettyMessagingService
Currently SO_SNDBUF & SO_RCVBUF used in `NettyMessagingService`  are not configurable.
The current default value is set to 1MB.

A user should be able to configure them using the following keys:
#### broker
-  zeebe.broker.network.so_sndbuf
- zeebe.broker.network.so_rcvbuf
#### gateway
-  zeebe.gateway.network.so_sndbuf
- zeebe.gateway.network.so_rcvbuf

The configs must be added to the `NetworkCfg` class (used by both broker & gateway), then mapped to `MessagingConfig`.

The docs must be updated as well to include these new configurations.

### PR DESCRIPTION
**Pull request by filipecampos**
## Description

This PR adds new configuration parameters to set the size of SO_SNDBUF and SO_RCVBUF.

## Checklist

Benchmarks
- [x] baseline (1MB)
- [x] 2MB
- [x] 3MB
- [x] 4MB
- [x] 6+4MB

Modified config unit tests
- [x] Broker
- [x] Gateway

## Related issues

closes #24104

### CODE REVIEW

File: zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/ConfigurationDefaults.java [hunk 1]
```diff
@@ -17,6 +17,12 @@ public final class ConfigurationDefaults {
   public static final String DEFAULT_CONTACT_POINT_HOST = "127.0.0.1";
   public static final int DEFAULT_CONTACT_POINT_PORT = 26502;
 
+  public static final int DEFAULT_SO_SNDBUF = 1024;
+  public static final int DEFAULT_SO_RCVBUF = 1024;
+  public static final int DEFAULT_BROKER_SO_SNDBUF = 1024;
```
[the remark this answers is not in the document]
Comment (filipecampos) [reply zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/ConfigurationDefaults.java]: Removed the broker defaults.

File: zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java [hunk 2]
```diff
@@ -250,6 +254,42 @@ public String getKeyStorePassword() {
...
+  public MessagingConfig setBrokerSo_Rcvbuf(final Integer brokerSo_Rcvbuf) {
+    this.brokerSo_Rcvbuf = brokerSo_Rcvbuf;
+    return this;
+  }
+
+  public Integer getGatewaySo_Sndbuf() {
+    return gatewaySo_Sndbuf;
+  }
+
+  public MessagingConfig setGatewaySo_Sndbuf(final Integer gatewaySo_Sndbuf) {
```
[the remark this answers is not in the document]
Comment (filipecampos) [reply zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java]: Removed the gateway methods.

File: zeebe/atomix/cluster/pom.xml [hunk 3]
```diff
@@ -272,6 +272,12 @@
       <artifactId>netty-resolver</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>org.springframework</groupId>
```
Comment (entangled90) [inline zeebe/atomix/cluster/pom.xml:276]: :x:  why did you need to add this? I don't think it's correct

File: zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java [hunk 4]
```diff
@@ -250,6 +253,24 @@ public String getKeyStorePassword() {
     return keyStorePassword;
   }
 
+  public DataSize getBrokerSoSndbuf() {
+    return brokerSoSndbuf;
+  }
+
+  public MessagingConfig setBrokerSoSndbuf(final DataSize brokerSoSndbuf) {
```
Comment (entangled90) [inline zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java:260]: :wrench:  why the field is prefixed with broker? Is this necessary?

File: zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java [hunk 5]
```diff
@@ -35,6 +36,8 @@ public class MessagingConfig implements Config {
   private CompressionAlgorithm compressionAlgorithm = CompressionAlgorithm.NONE;
   private File keyStore;
   private String keyStorePassword;
+  private DataSize brokerSoSndbuf = DataSize.ofMegabytes(1);
```
Comment (entangled90) [inline zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java:39]: :question:  is this default needed? I see that you added default values in Broker config. I would avoid having too many defaults in different parts of the code base

File: zeebe/gateway/src/test/java/io/camunda/zeebe/gateway/impl/configuration/GatewayCfgTest.java [hunk 6]
```diff
@@ -174,13 +176,17 @@ public void shouldUseEnvironmentVariables() {
     setEnv("zeebe.gateway.filters.0.id", "overwrittenFilter");
     setEnv("zeebe.gateway.filters.0.className", "OverwrittenFilter");
     setEnv("zeebe.gateway.filters.0.jarPath", "./overwrittenFilter.jar");
+    setEnv("zeebe.gateway.network.soSndbuf", DEFAULT_GATEWAY_SO_SNDBUF.toString());
```
Comment (entangled90) [inline zeebe/gateway/src/test/java/io/camunda/zeebe/gateway/impl/configuration/GatewayCfgTest.java:179]: :x: why do you set the environment to the default values?
The value provided in the env must be different than the default for the test to represent a usecase

File: zeebe/atomix/cluster/pom.xml (hunk 3 above)
Comment (filipecampos) [reply zeebe/atomix/cluster/pom.xml:276]: Since we're using `DataSize` in `MessagingConfig` it became a dependency caught by the build validations.

File: zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java (hunk 5 above)
Comment (filipecampos) [reply zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java:39]: It prevents the NPE that occurs during Broker startup as we're retrieving the new parameters values and they're null which oddly seems to occurs on the second time `bootstrapServer` is invoked.

Comment (entangled90) [reply zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java:39]: Does it happen in unit tests or in qa tests?

Comment (filipecampos) [reply zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/MessagingConfig.java:39]: In alternative, it seems we can set the value of the parameters in the other instance of `MessagingConfig`.

**Comment by filipecampos** [comment]
### Benchmarks

<p>Ran benchmarks for the following values in <em>socketReceiveBuffer</em> and <em>socketSendBuffer</em> both for Broker and Gateway:</p>

|  | **Broker** |  | **Gateway** |  | **Grafana** |
| --- | --- | --- | --- | --- | --- |
|  | **socketReceiveBuffer** | **socketSendBuffer** | **socketReceiveBuffer** | **socketSendBuffer** |  |
| **1MB - Baseline** | 1MB | 1MB | 1MB | 1MB | https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-1a&var-pod=All&var-partition=All&from=1732799460000&to=1732803060000 |
| **2MB** | 2MB | 2MB | 2MB | 2MB | https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-2&var-pod=All&var-partition=All&from=1732790880000&to=1732794480000 |
| **3MB** | 3MB | 3MB | 3MB | 3MB | [https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-3b&var-pod=All&var-partition=All&from=1732801920000&to=1732805220000](https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-3b&var-pod=All&var-partition=All&from=1732801920000&to=1732805520000) |
| **4MB** | 4MB | 4MB | 4MB | 4MB | https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-4&var-pod=All&var-partition=All&from=1732791000000&to=1732794600000 |
| **6+4MB** | 6MB | 4MB | 6MB | 4MB | [https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-6a&var-pod=All&var-partition=All&from=1732800900000&to=1732804800000](https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-6a&var-pod=All&var-partition=All&from=1732800900000&to=1732804500000) |


<p>We used the values-realistic-benchmark.yaml file and just changed <em>zeebe.resources.limits.cpu</em> to 2500m.</p>
<p>Used the following command to check the configurations set on the pod. The presented output is for the 4+6MB scenario.</p>
<pre><code class="language-jsx">$ k describe configmap benchmark-config
Name:         benchmark-config
Namespace:    fc-socket-params-6a
Labels:       app.kubernetes.io/instance=fc-socket-params-6a
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=zeebe-benchmark
              app.kubernetes.io/version=8.2.5
              helm.sh/chart=zeebe-benchmark-0.3.8
Annotations:  meta.helm.sh/release-name: fc-socket-params-6a
              meta.helm.sh/release-namespace: fc-socket-params-6a

Data
====
benchmark-config.yaml:
----
zeebe:
  broker:
    network:
      socketReceiveBuffer: 6MB
      socketSendBuffer: 4MB
  gateway:
    network:
      socketReceiveBuffer: 6MB
      socketSendBuffer: 4MB
zeebe.broker.data.diskUsageCommandWatermark: &quot;0.8&quot;
zeebe.broker.data.diskUsageReplicationWatermark: &quot;0.9&quot;
zeebe.broker.executionMetricsExporterEnabled: &quot;true&quot;
zeebe.broker.experimental.consistencyChecks.enableForeignKeyChecks: &quot;true&quot;
zeebe.broker.experimental.consistencyChecks.enablePreconditions: &quot;true&quot;
zeebe.broker.exporters.elasticsearch.args.index.indexSuffixDatePattern: yyyy-MM-dd_HH
zeebe.broker.flowControl.write.enabled: &quot;true&quot;
zeebe.broker.flowControl.write.limit: 4000
zeebe.gateway.monitoring.enabled: &quot;true&quot;
zeebe.gateway.threads.managementThreads: &quot;1&quot;

BinaryData
====

Events:  &lt;none&gt;

</code></pre>



### Results
There aren’t any significant differences in the metrics by looking at the various Grafana dashboards, even when looking at:

- Throughput
- Latencies
- CPU usage
- Memory usage
- Disk usage

The only noticeable differences are:
- the processing queue for the various partitions in the 1MB scenario are very close to 0 with some rare spikes up to 200-300. In the 2MB scenario, only one partition is frequently closer to 0 and the other two partitions vary a lot. We see similar results to 1MB in the remaining scenarios but with more frequent spikes.
- the amount of data transmitter per second through the network which seems to increase with the size of the sending buffer as we can see in the following charts.
image
1MB

image
4MB


- the commit latency on the 1MB scenario seems to be more stable than in the remaining scenarios with p99 more than 90% below 25ms and not going over 35ms whereas on the other scenarios some p99 reach over 150ms

**Comment by entangled90** [comment]
Looks like 1MB is slightly better than the others in some metrics. Should we try 256KB &  512KB?

**Review by entangled90 — CHANGES_REQUESTED** [review]
Great, just fix the test cases related to reading from environment variables

File: zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/configuration/BrokerCfgTest.java [hunk 7]
```diff
@@ -528,6 +532,47 @@ public void shouldSetCustomMembershipConfig() {
...
+    final BrokerCfg cfg = TestConfigReader.readConfig("cluster-cfg", environment);
+
+    // then
+    assertThat(cfg.getNetwork().getSocketSendBuffer().toBytes()).isEqualTo(2097152);
+  }
+
+  @Test
+  public void shouldUseSoRcvbufFromEnvironment() {
+    environment.put(ZEEBE_BROKER_NETWORK_SOCKET_RECEIVE_BUFFER, "3MB");
+    assertSocketBufferParameter(ZEEBE_BROKER_NETWORK_SOCKET_RECEIVE_BUFFER, "3MB");
```
[the remark this answers is not in the document]
Comment (filipecampos) [reply zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/configuration/BrokerCfgTest.java]: Changed the test.

File: zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/configuration/BrokerCfgTest.java [hunk 8]
```diff
@@ -528,6 +532,47 @@ public void shouldSetCustomMembershipConfig() {
...
+    assertSocketBufferParameter(ZEEBE_BROKER_NETWORK_SOCKET_SEND_BUFFER, "2MB");
+  }
+
+  private void assertSocketBufferParameter(
+      final String socketBufferParameter, final String number) {
+    assertThat(environment.get(socketBufferParameter)).isEqualTo(number);
+  }
+
+  @Test
+  public void shouldOverrideSpecifiedSoSndbufFromEnvironment() {
```
[the remark this answers is not in the document]
Comment (filipecampos) [reply zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/configuration/BrokerCfgTest.java]: Renamed.

File: zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/configuration/BrokerCfgTest.java [hunk 9]
```diff
@@ -528,6 +532,57 @@ public void shouldSetCustomMembershipConfig() {
...
+    // given no value is defined
+
+    // when
+    final BrokerCfg cfg = TestConfigReader.readConfig("cluster-cfg", environment);
+
+    // then
+    assertThat(cfg.getNetwork().getSocketSendBuffer().toKilobytes()).isEqualTo(1024);
+  }
+
+  private void assertSocketBufferParameter(
```
Comment (entangled90) [inline zeebe/broker/src/test/java/io/camunda/zeebe/broker/system/configuration/BrokerCfgTest.java:546]: :thought_balloon:  is this method used anywhere?

**Review by entangled90 — APPROVED** [review]
:rocket:  just check my comment before merging

**Comment by filipecampos** [comment]
> Veryt good analysis! Looks like 1MB is slightly better than the others in some metrics. Not sure if it's worth trying 256KB & 512KB as the gain is probably marginal (and could likely depend on the type of benchmark we run)

|  | **Broker** |  | **Gateway** |  | **Grafana** |
| --- | --- | --- | --- | --- | --- |
|  | **socketReceiveBuffer** | **socketSendBuffer** | **socketReceiveBuffer** | **socketSendBuffer** |  |
| **256KB** | 256KB | 256KB | 256KB | 256KB | https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-256&var-pod=All&var-partition=All&from=1733142600000&to=1733146200000 |
| **512KB** | 512KB | 512KB | 512KB | 512KB | https://grafana.dev.zeebe.io/d/zeebe-dashboard/zeebe?orgId=1&var-DS_PROMETHEUS=prometheus&var-cluster=All&var-namespace=fc-socket-params-512&var-pod=All&var-partition=All&from=1733156400000&to=1733160000000 |

**256 and 512KB Analysis**

As with the previous benchmarks, there aren’t any significant differences in the metrics by looking at the various Grafana dashboards.

Compared to the baseline, both 256 and 512 scenarios had some records that weren’t exported nor processed and the processing queue size was often above the 100 and 150 limits. In 256, the large majority of those cases seems to be due to Partition 2 whereas on 512 the occurrences were split by various partitions. We also see higher request-response latency values than in the baseline, which seems to improve with the increase in the buffers size.

**Comment by filipecampos** [comment]
Documentation will be updated in https://github.com/camunda/camunda-docs/issues/4694
